### ISSUE TITLE
G Suite group membership is not expanded for certain users

### ISSUE THREAD

**Issue by rich-boyce** [forseti-security/forseti-security#2969]
- Forseti version: 2.17.0
 - Module: explainer
 - No relevant errors are presented in either the client or server logs.
 - Deployed via Deployment Manager following https://forsetisecurity.org/docs/latest/setup/install.html and https://forsetisecurity.org/docs/latest/configure/inventory/gsuite.html.
 - Reporting this bug as requested in the Slack channel by @joecheuk

We're seeing an issue where Forseti apparently is unaware that some G Suite users are members of groups. Note that other users' group memberships _are_ properly expanded, so G Suite integration is working at least to some level.

For example:

- user/rich@example.org and user/tom@example.org are both members of group/ops@example.org
- group/ops@example.org has a large number of effective permissions:
`rich@forseti-client-vm-8fe5e30:~$ forseti explainer access_by_member group/ops@example.org | grep role | wc -l
243`
- user/rich@example.org is shown to inherit all these permissions:
`rich@forseti-client-vm-8fe5e30:~$ forseti explainer access_by_member user/rich@example.org | grep role | wc -l
243`
- user/tom@example.org is *not* shown to inherit them:
`rich@forseti-client-vm-8fe5e30:~$ forseti explainer access_by_member user/tom@example.org | grep role | wc -l
8`
- We have verified that, in GCP, user/tom@example.org _can_ perform actions using permissions given by his membership of group/ops@example.org, including actions that Forseti reports he should _not_ be able to, for example he can in fact power off this instance:
`rich@forseti-client-vm-8fe5e30:~$ forseti explainer check_policy instance/xxx compute.instances.stop user/tom@example.org
{
  "result": false
}`
- There are several users in our organisation affected by this issue. A cursory check seems to imply that users of long standing are affected; new joiners are not. An exhaustive test of this theory has not been completed as yet.
- For an affected user, the roles listed by `forseti explainer access_by_member` for that user are _only_ those granted directly to the user, not to a group.
- For unaffected users, _both_ direct user permissions _and_ group permissions are listed by `forseti explainer access_by_member`

So **it seems that we have a subset of users for which Forseti is unable to evaluate group memberships**.

Can someone please assist with this issue?

**Comment by rich-boyce** [comment]
Hi all, any idea when this issue will get any attention? It's blocking a primary use case we have for Forseti.

Thanks.

**Comment by blueandgold** [comment]
I looked into this, and can repro the problem.  Digging deeper, I am able to find a bug, where not all the group members are not fully inventoried.  i.e. some members are not listed in the group.

For example:
* Suppose we have Group `XYZ`, with members `111`, `222`, `333` in the actual Google Groups.
* I expect members `111`, `222`, `333` to be members of Group `XYZ` in the Forseti `gcp_inventory` table.
* However, I am only able to see partial members, such as only `111`, `333` in Group `XYZ`, and `222` is missing.

The bug seems to be [in this code](https://github.com/forseti-security/forseti-security/blob/dev/google/cloud/forseti/services/inventory/storage.py#L1341-L1374)

![image](https://user-images.githubusercontent.com/6977544/62104448-829f8480-b254-11e9-91e6-a8d5058cbc0d.png)

Basically, if `222` is a member of group `ABC`, then an update will be triggered for `222` in Group `ABC`, instead of a new insert for Group `XYZ`.

I will take a look at tomorrow to see how to fix this.

**Comment by blueandgold** [comment]
I also checksummed all the resources by type in the DB.
